File: singly_linked_list/singly_linked_list.py
class ListNode:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next

    def get_next(self):
        return self.next

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_to_head(self, value):
        node = ListNode(value)

        if self.head is not None:
            node.next = self.head
        else:
            self.tail = node

        self.head = node

    def add_to_tail(self, value):
        node = ListNode(value)

        if self.tail is not None:
            self.tail.next = node
        else:
            self.head = node

        self.tail = node

    def contains(self, value):
        current = self.head

        while current is not None:
            if current.value == value:
                return True
            current = current.next
        return False

    def remove_head(self):
        if self.head is None:
            return None

        if self.head == self.tail:
            self.tail = None

        old_head = self.head
        new_head = self.head.next
        self.head.next = None
        self.head = new_head

        return old_head.value

    def get_max(self):
        if self.head is None:
            return None

        max = 0
        current = self.head

        while current is not None:
            if current.value > max:
                max = current.value
            current = current.next

        return max

    def reverse_list(self, head, prev=None):
        prev = prev
        curr = head
        next = None

        while curr is not None:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next

        self.head = prev

File: data_structures/singly_linked_list/singlyLinkedList.ts
interface Node {
  value: any;
  next: Node;
}

interface SLL {
  head: Node;
  tail: Node;
}

export class ListNode implements Node {
  constructor(public value: any, public next: Node = null) {}
}

export class LinkedList implements SLL {
  constructor(public head: Node = null, public tail: Node = null) {}

  addToHead(value) {
    const node = new ListNode(value);

    if (this.head !== null) {
      node.next = this.head;
    } else {
      this.tail = node;
    }

    this.head = node;
  }

  addToTail(value) {
    const node = new ListNode(value);

    if (this.tail !== null) {
      this.tail.next = node;
    } else {
      this.head = node;
    }

    this.tail = node;
  }

  contains(value) {
    if (this.head !== null) {
      let current = this.head;
      while (current !== null) {
        if (current.value === value) {
          return true;
        }
        current = current.next;
      }
    }
    return false;
  }

  removeHead() {
    if (this.head === null) {
      return null;
    }

    if (this.head === this.tail) {
      this.tail = null;
    }

    const oldHead = this.head;
    const newHead = this.head.next;
    this.head.next = null;
    this.head = newHead;

    return oldHead.value;
  }

  getMax() {
    if (this.head === null) {
      return null;
    }

    let max = 0;
    let current = this.head;

    while (current !== null) {
      if (current.value > max) {
        max = current.value;
      }
      current = current.next;
    }

    return max;
  }

  reverseList(head, previous = null) {
    let prev = previous;
    let curr = head;
    let next = null;

    while (curr !== null) {
      next = curr.next;
      curr.next = prev;
      prev = curr;
      curr = next;
    }

    this.head = prev;
  }
}
